### PR TITLE
SAA-564 service and domain changes to support automatic releasing an offender from prison.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
Service layer and Domain changes to support an offender being released prison (events).

The incoming event work to call this will be done separately.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/CapacityService.kt [hunk 1]
```diff
@@ -58,6 +58,6 @@ class CapacityService(
   }
 
   private fun ActivitySchedule.getAllocationsForToday(): List<Allocation> {
-    return this.allocations().filter { it.status(PrisonerStatus.ACTIVE) || it.status(PrisonerStatus.SUSPENDED) }
+    return this.allocations().filterNot { it.status(PrisonerStatus.ENDED) }
   }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/CapacityService.kt:62]: Like it

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderEventsOfInterestService.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service
+
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.entity.Allocation
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.entity.PrisonerStatus
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.repository.AllocationRepository
+import java.time.LocalDateTime
+
+@Service
+class OffenderEventsOfInterestService(private val repository: AllocationRepository) {
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderEventsOfInterestService.kt:10]: I know what you mean on the name. Maybe a name `InboundEventsService` is clearer? Their could be a method within it which determines whether its interesting or not? i.e. relates to somebody with an allocation that could be affected by this event. Fine for now .. just thinking ahead.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderEventsOfInterestService.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    }.saveAffectedAllocations()
+
+  private fun List<Allocation>.deallocateAndSaveAffectedAllocations() =
+    LocalDateTime.now().let { now ->
+      this.filterNot { it.status(PrisonerStatus.ENDED) }.map { it.deallocate(now, "Released from prison") }
+    }.saveAffectedAllocations()
+
+  private fun List<Allocation>.saveAffectedAllocations() =
+    repository.saveAllAndFlush(this).toList()
+}
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderEventsOfInterestService.kt:32]: Very neat and tidy!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderReleasedFromPrisonServiceTest.kt [hunk 4]
```diff
@@ -11,26 +12,40 @@ import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.entity.Prisoner
...
+    whenever(repository.findByPrisonCodeAndPrisonerNumber(moorlandPrisonCode, "123456")).thenReturn(
+      previouslyActiveAllocations,
+    )
 
     service.temporarilyReleaseOffender(moorlandPrisonCode, "123456")
 
-    allocations.forEach { assertThat(it.status(PrisonerStatus.AUTO_SUSPENDED)).isTrue }
+    previouslyActiveAllocations.forEach {
+      assertThat(it.status(PrisonerStatus.AUTO_SUSPENDED)).isTrue
+      assertThat(it.suspendedBy).isEqualTo("SYSTEM")
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderReleasedFromPrisonServiceTest.kt:45]: Is Mike aware of the new allocation status AUTO_SUSPENDED?

**Review by timharrison-moj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderEventsOfInterestService.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderEventsOfInterestService.kt:10]: Yeah I like your suggestion better.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderReleasedFromPrisonServiceTest.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/OffenderReleasedFromPrisonServiceTest.kt:45]: Nope not yet.  @mikehalmamoj incoming soon!!

**Pull request merged**